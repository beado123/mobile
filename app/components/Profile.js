import React, {Component} from 'react';
import { StackNavigator } from 'react-navigation';
import { Buffer } from 'buffer'
import { StyleSheet,Text,View,ScrollView,Image,AppRegistry,Button,ImageBackground,TouchableOpacity, AsyncStorage, RefreshControl} from 'react-native';
import styleHeader from '/home/yihan/AwesomeProject/app/styles/Header';
import styleBar from '/home/yihan/AwesomeProject/app/styles/Bar';

{/*Design follows from https://www.youtube.com/watch?v=hFHMboJk6yg&t=474s*/}
export default class ProfilePage extends React.Component {

    state = {
        login: '',
        username: '',
        followers: -1,
        following: -1,
        public_repos: -1,
        avatar_url: 'blah',
        created_at: '',
        refreshing: false,
    };

    componentWillMount(){
        this.fetchData();
    };

    /*This function fetches data from github api.  */
    fetchDataFromAPI =  async (login, ending, username, password) => {
        const response = await fetch('http://api.github.com/user' + ending, {
            method: 'GET',
            headers: {
                'Authorization': 'Basic ' + new Buffer(username + ':' + password).toString('base64')
            },
        });
        let responseJson = await response.json();
        return responseJson;
    };

    /*This function sets up the data needed when loading the profile view. */
    fetchData = async () => {
        const username = await AsyncStorage.getItem('username');
        const password = await AsyncStorage.getItem('password');
        const { params } = this.props.navigation.state;
        const login = params ? params.name : username;
        const userJson = await this.fetchDataFromAPI(login, '', username, password);
        const followingJson = await this.fetchDataFromAPI(login, '/following', username, password);
        const followersJson = await this.fetchDataFromAPI(login, '/followers', username, password);
        const reposJson = await this.fetchDataFromAPI(login, '/repos', username, password);
        if( login == username){
            AsyncStorage.setItem('followers', JSON.stringify(followersJson));
            AsyncStorage.setItem('following', JSON.stringify(followingJson));
            AsyncStorage.setItem('repos', JSON.stringify(reposJson));
        }
        this.setState({followers: userJson.followers,
                       following: userJson.following,
                       public_repos: userJson.public_repos,
                       login: login,
                       username: username,
                       avatar_url: userJson.avatar_url,
                       created_at: userJson.created_at.substring(0,9)
        });
    };

    /*This function updates number of following, followers on the profile page. */
    updateData = async () => {
        let followingJson = await AsyncStorage.getItem('following');
        let following = JSON.parse(followingJson);
        let followersJson = await AsyncStorage.getItem('followers');
        let followers = JSON.parse(followersJson);
        this.setState({followers: followers.length, following: following.length});
    };

    /*This function handles refreshing of current list. */
    _onRefresh = () => {
        this.setState({refreshing: true});
        this.updateData().then(() => {
            this.setState({refreshing: false});
        });
    }

  render(){
    return (
        <ScrollView refreshControl={
                      <RefreshControl
                        refreshing={this.state.refreshing}
                        onRefresh={this._onRefresh.bind(this)}
                      />
                        }contentContainerStyle={{flexGrow: 1,}} scrollEnabled>

            <View style={{flex:1,flexDirection: 'column'}}  >
                <ImageBackground style={styleHeader.headerBackground} source={require('/home/yihan/AwesomeProject/app/img/wallpaper2.jpg')}>
                    <View style={styleHeader.header}>
                        <View style={styleHeader.profilepicWrap}>
                            <Image style={styleHeader.profilepic} source={{uri: this.state.avatar_url}} />
                        </View>
                        <Text style={styleHeader.name}>{this.state.login}</Text>
                        {this.state.login != this.state.username && <TouchableOpacity style={styleHeader.button} onPress = {() =>this.props.navigation.goBack()}>
                            <Text >Go Back</Text>
                        </TouchableOpacity>}
                        {this.state.login == this.state.username && <TouchableOpacity style={styleHeader.buttonWide} onPress = {() =>this.props.navigation.navigate('notifications',{name: this.state.login})}>
                            <Text >Notifications</Text>
                        </TouchableOpacity>}
                        {this.state.login == this.state.username && <TouchableOpacity style={styleHeader.buttonWide} onPress = {() =>this.props.navigation.navigate('visualization')}>
                            <Text >Visualization</Text>
                        </TouchableOpacity>}
                    </View>

                    <View style={styleBar.bar}>
                        <View style={styleBar.firstBar}>
                          <View style={[styleBar.barItem, styleBar.barseparator]}>
                            <Text style={styleBar.barTop}>{this.state.followers}</Text>
                            <Text style={styleBar.barBottom} onPress={() => this.props.navigation.navigate('followers',{name: this.state.login})}>Followers</Text>
                          </View>
                          <View style={[styleBar.barItem, styleBar.barseparator]}>
                            <Text style={styleBar.barTop}>{this.state.following}</Text>
                            <Text style={styleBar.barBottom} onPress={() => this.props.navigation.navigate('following',{name: this.state.login})}>Following</Text>
                          </View>
                          <View style={styleBar.barItem}>
                            <Text style={styleBar.barTop}>{this.state.public_repos}</Text>
                            <Text style={styleBar.barBottom} onPress={() => this.props.navigation.navigate('repository',{name: this.state.login})}>Repository</Text>
                          </View>
                        </View>
                        <View style={styleBar.dateBar}>
                            <Text style={styleBar.barTop}>Created at</Text>
                            <Text style={styleBar.barBottom}>{this.state.created_at}</Text>
                        </View>
                    </View>
                </ImageBackground>
            </View>
      </ScrollView>
    );
  }
}
