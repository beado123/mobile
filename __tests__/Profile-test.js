import React from 'react';
import "isomorphic-fetch"
import Profile from '/home/yihan/AwesomeProject/app/components/Profile.js';
import renderer from 'react-test-renderer';

var num_followers = -1;
var num_following = -1;
test('renders correctly', () => {
  const tree = renderer.create(<Profile />).toJSON();
  expect(tree).toMatchSnapshot();
});

test('the user data is of correct type', async () => {
    const XMLHttpRequest = require('xhr2');
    global.XMLHttpRequest = XMLHttpRequest;
    let profileComponent = renderer.create(<Profile/>).getInstance();
    return profileComponent.fetchDataFromAPI('beado123','', 'beado123', '62524391a/').then(data => {
        expect(data.type).toBe('User');
        num_followers = data.followers;
        num_following = data.following;
    });
});

test('get number of following correctly', async () => {
    const XMLHttpRequest = require('xhr2');
    global.XMLHttpRequest = XMLHttpRequest;
    let profileComponent = renderer.create(<Profile/>).getInstance();
    return profileComponent.fetchDataFromAPI('beado123','/following', 'beado123', '62524391a/').then(data => {
        expect(data.length).toBe(num_following);
    });
});

test('get number of followers correctly', async () => {
    const XMLHttpRequest = require('xhr2');
    global.XMLHttpRequest = XMLHttpRequest;
    let profileComponent = renderer.create(<Profile/>).getInstance();
    return profileComponent.fetchDataFromAPI('beado123','/followers', 'beado123', '62524391a/').then(data => {
        expect(data.length).toBe(num_followers);
    });
});

test('refreshing state should change', () => {
    let profileComponent = renderer.create(<Profile/>).getInstance();
    profileComponent._onRefresh();
    expect(profileComponent.state.refreshing).toBe(true);
});
