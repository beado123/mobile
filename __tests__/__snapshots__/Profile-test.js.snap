// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  scrollEnabled={true}
>
  <RCTRefreshControl />
  <View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          Object {
            "alignSelf": "stretch",
            "flex": 1,
          }
        }
      >
        <Image
          source={1}
          style={
            Array [
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              },
              Object {
                "height": undefined,
                "width": undefined,
              },
              undefined,
            ]
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "rgba(0,0,0,0.5)",
              "flex": 10,
              "justifyContent": "center",
              "padding": 10,
            }
          }
        >
          <View
            style={
              Object {
                "borderColor": "#000",
                "borderRadius": 30,
                "borderWidth": 1,
                "height": 90,
                "width": 90,
              }
            }
          >
            <Image
              source={
                Object {
                  "uri": "blah",
                }
              }
              style={
                Object {
                  "alignItems": "stretch",
                  "borderColor": "#cfd2d4",
                  "borderWidth": 5,
                  "flex": 1,
                  "width": null,
                }
              }
            />
          </View>
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#fff",
                "fontSize": 16,
                "fontWeight": "bold",
                "marginTop": 20,
              }
            }
          >
            
          </Text>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#BFDDEF",
                "borderColor": "#000",
                "borderRadius": 30,
                "height": 50,
                "justifyContent": "center",
                "marginTop": 10,
                "opacity": 1,
                "width": 90,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              Notifications
            </Text>
          </View>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#BFDDEF",
                "borderColor": "#000",
                "borderRadius": 30,
                "height": 50,
                "justifyContent": "center",
                "marginTop": 10,
                "opacity": 1,
                "width": 90,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              Visualization
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#64678f",
              "flex": 0.4,
              "flexDirection": "column",
            }
          }
        >
          <View
            style={
              Object {
                "borderTopColor": "#000",
                "borderTopWidth": 4,
                "flex": 0.3,
                "flexDirection": "row",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "padding": 15,
                  },
                  Object {
                    "borderRightWidth": 4,
                  },
                ]
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#fff",
                    "fontSize": 12,
                    "fontStyle": "italic",
                    "fontWeight": "bold",
                  }
                }
              >
                -1
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                onPress={[Function]}
                style={
                  Object {
                    "color": "#fff",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "textAlign": "center",
                    "width": 80,
                  }
                }
              >
                Followers
              </Text>
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flex": 1,
                    "padding": 15,
                  },
                  Object {
                    "borderRightWidth": 4,
                  },
                ]
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#fff",
                    "fontSize": 12,
                    "fontStyle": "italic",
                    "fontWeight": "bold",
                  }
                }
              >
                -1
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                onPress={[Function]}
                style={
                  Object {
                    "color": "#fff",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "textAlign": "center",
                    "width": 80,
                  }
                }
              >
                Following
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "padding": 15,
                }
              }
            >
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Object {
                    "color": "#fff",
                    "fontSize": 12,
                    "fontStyle": "italic",
                    "fontWeight": "bold",
                  }
                }
              >
                -1
              </Text>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                onPress={[Function]}
                style={
                  Object {
                    "color": "#fff",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "textAlign": "center",
                    "width": 80,
                  }
                }
              >
                Repository
              </Text>
            </View>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "borderTopColor": "#000",
                "borderTopWidth": 4,
                "flex": 0.1,
                "padding": 10,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#fff",
                  "fontSize": 12,
                  "fontStyle": "italic",
                  "fontWeight": "bold",
                }
              }
            >
              Created at
            </Text>
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#fff",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "textAlign": "center",
                  "width": 80,
                }
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
